### PR TITLE
Fix refund restock and shipment issues (SHUUP-2915, SHUUP-2917, SHUUP-2958)

### PR DESCRIPTION
**Pull request by shawnadelic**
- Fix payment and shipment dropdown visibility in Order admin
- Fix stock management refund issues
- Fix shipment creation issues for refunded orders

Refs SHUUP-2915, SHUUP-2917, SHUUP-2958

### CODE REVIEW

File: shuup/admin/modules/orders/views/detail.py [hunk 1]
```diff
@@ -33,15 +33,15 @@ def get_toolbar(self):
         order = self.object
         toolbar = Toolbar()
         action_menu_items = []
-        if not (order.is_paid() or order.is_canceled() or order.has_refunds()):
+        if not (order.is_paid() or order.is_canceled() or (order.has_refunds() and not order.can_create_refund())):
```
Comment (tulimaki) [inline shuup/admin/modules/orders/views/detail.py]: Could this be can_create_payment-method for order?

File: shuup/admin/modules/orders/views/detail.py [hunk 2]
```diff
@@ -33,15 +33,15 @@ def get_toolbar(self):
...
             action_menu_items.append(
                 DropdownItem(
                     url=reverse("shuup_admin:order.create-payment", kwargs={"pk": order.pk}),
                     icon="fa fa-money",
                     text=_("Create Payment"),
                 )
             )
-        if (order.get_unshipped_products() and not order.is_canceled() and not order.has_refunds()):
+        if (order.get_unshipped_products() and not order.is_canceled()):
```
Comment (tulimaki) [inline shuup/admin/modules/orders/views/detail.py]: Could this be can_create_shipment-method for order?

File: shuup/core/models/_order_lines.py [hunk 3]
```diff
@@ -93,6 +94,10 @@ class OrderLine(MoneyPropped, models.Model, Priceful):
     base_unit_price = PriceProperty('base_unit_price_value', 'order.currency', 'order.prices_include_tax')
     discount_amount = PriceProperty('discount_amount_value', 'order.currency', 'order.prices_include_tax')
 
+    # The following fields apply to product refund lines
+    refunded_quantity = QuantityField(blank=True, null=True, verbose_name=_('refunded_quantity'))
```
Comment (tulimaki) [inline shuup/core/models/_order_lines.py]: Is this field really needed. Can we get the quantity based on linetype.

File: shuup/core/models/_order_lines.py [hunk 4]
```diff
@@ -93,6 +94,10 @@ class OrderLine(MoneyPropped, models.Model, Priceful):
...
+    restocked_quantity = QuantityField(blank=True, null=True, verbose_name=_('restocked_quantity'))
```
Comment (tulimaki) [inline shuup/core/models/_order_lines.py]: orderline shouldn't probably know anything about stocks.

File: shuup/core/suppliers/base.py [hunk 5]
```diff
@@ -7,11 +7,21 @@
 # LICENSE file in the root directory of this source tree.
 from django.core.exceptions import ValidationError
 from django.utils.translation import ugettext_lazy as _
+from enumfields import Enum
 
 from shuup.core.models import StockBehavior
 from shuup.core.stocks import ProductStockStatus
 
 
+class StockAdjustmentType(Enum):
```
Comment (tulimaki) [inline shuup/core/suppliers/base.py]: You could probably move this under suppliers/enums.py or something like that to dodge import problems.

File: shuup/core/models/_orders.py [hunk 6]
```diff
@@ -558,6 +558,18 @@ def create_refund(self, refund_data, created_by=None):
             if amount > self.taxful_total_price.amount:
                 raise RefundExceedsAmountException
 
+            # If restocking products, calculate quantity of products to restock
+            restock_quantity = 0
+            if parent_line.type == OrderLineType.PRODUCT:
```
Comment (shawnadelic) [inline shuup/core/models/_orders.py]: This part should be more readable

File: shuup/core/models/_orders.py [hunk 7]
```diff
@@ -558,17 +567,31 @@ def create_refund(self, refund_data, created_by=None):
             if amount > self.taxful_total_price.amount:
                 raise RefundExceedsAmountException
 
+            # If restocking products, calculate quantity of products to restock
+            product = parent_line.product
+            if (restock_products and quantity and product and (product.stock_behavior == StockBehavior.STOCKED)):
+                from shuup.core.suppliers.enums import StockAdjustmentType
```
Comment (shawnadelic) [inline shuup/core/models/_orders.py]: Note: I was still getting a circular import error here, so this seems like it might be necessary

**Pull request merged**